### ISSUE TITLE
Can you execute raw queries from code?

### ISSUE THREAD

**Issue by abrull** [litedb-org/LiteDB#2010]
I want to execute raw sql queries from code. For example I would like to export a collection to a json file. I know you can do that in LiteDb.Studio but what about c#?

**Comment by akashchopra** [comment]
Try the `Execute` method on (I)LiteDatabase: https://github.com/mbdavid/LiteDB/blob/master/LiteDB/Client/Database/LiteDatabase.cs#L243

**Comment by isaacabraham** [comment]
@akashchopra this returns an `IBsonDataReader`. How do you convert that into the collection type?

**Comment by isaacabraham** [comment]
@akashchopra I don't see why it would be a big task - you would indicate the return type with a generic type argument. This is how most such tools work e.g. Dapper has had this for years.

**Comment by grofit** [comment]
Sorry to jump in on this but it would be good if the docs for `Execute(sql, params)` could be updated to indicate how you are supposed to use params, i.e do you just put `Execute("SELECT $ FROM a WHERE $.b = @v1", new { v1 = "value"});` or something? its not clear as to how the variable replacement mechanism works.

**Comment by JKamsker** [comment]
Yes. `LiteDatabase.Execute` runs the same SQL you use in LiteDB.Studio. It returns an `IBsonDataReader`, which you can iterate or turn into an `IEnumerable<BsonValue>` with `.ToEnumerable()`.

There are two ways to pass parameters:
- **named**: `@name`, looked up in a `BsonDocument`
- **positional**: `@0`, `@1`, … from a `params BsonValue[]`

Anonymous objects aren't accepted directly, but `db.Mapper.ToDocument(new { … })` turns one into the parameter document. There is no `Execute<T>`. To get typed results, map each row with the mapper:

```csharp
// named
using var reader = db.Execute("SELECT $ FROM customers WHERE Age > @minAge",
    new BsonDocument { ["minAge"] = 40 });

// positional
var rows = db.Execute("SELECT $ FROM customers WHERE Name = @0 OR Name = @1", "Ann", "Cid")
    .ToEnumerable();

// anonymous object -> parameter document
var p = db.Mapper.ToDocument(new { minAge = 40 });
var count = db.Execute("SELECT $ FROM customers WHERE Age > @minAge", p).ToEnumerable().Count();

// typed results
var customers = db.Execute("SELECT $ FROM customers WHERE Age > @0", 40)
    .ToEnumerable()
    .Select(d => db.Mapper.ToObject<Customer>(d.AsDocument))
    .ToList();
```

To export a collection to JSON, you can use SQL, the same way you would in Studio:

```csharp
db.Execute("SELECT $ INTO $file('C:/temp/customers.json') FROM customers");
```

Or you can do it in code: `JsonSerializer.Serialize(new BsonArray(db.Execute("SELECT $ FROM customers").ToEnumerable()))`.

For typed queries in regular code, `col.Query().Where(...).ToList()` or `col.Find(x => ...)` is usually simpler than SQL. `Execute` is most useful for ad-hoc commands like the export above. This works the same on 5.0.21 and the 6.x prerelease.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
